README.md
# technixo_k_chart

[![License](https://img.shields.io/badge/license-MIT-green.svg)](/LICENSE)

## 介绍
一个仿火币的flutter图表库包含深度图，支持各种指标及放大缩小、平移等操作

[webdemo演示](https://flutter-widget.live/packages/flutter_k_chart)
Demo v0.1.0：[下载 APK](https://github.com/415593725/flutter_k_chart/blob/master/k_chart.apk)

## 演示

![chart_imge](https://github.com/gwhcn/flutter_k_chart/blob/master/example/images/k_chart.2019-09-01%202010_19_56.gif)
![depth_image](https://github.com/gwhcn/flutter_k_chart/blob/master/example/images/depth.2019-09-01%202010_21_31.gif)
![image1](https://github.com/gwhcn/flutter_k_chart/blob/master/example/images/screenshots.png)

## 简单用例
#### 1.在 pubspec.yaml 中添加依赖
本项目数据来自火币openApi，火币的接口可能需要翻墙，接口失败后会加载本地json。由于项目没有很好的封装，建议使用本地方式使用
```yaml
//本地导入方式
dependencies:
  technixo_k_chart:
    path: 项目路径
```

#### 2.在布局文件中添加
```dart
import 'package:technixo_k_chart/technixo_k_chart.dart';
....
Container(
  height: 450,
  width: double.infinity,
  child: KChartWidget(
    datas,//数据
    isLine: isLine,//是否显示折线图
    mainState: _mainState,//控制主视图指标线
    secondaryState: _secondaryState,//控制副视图指标线
    volHidden: true,// hide or show vol chart
    fractionDigits: 4,//保留小数位数
  ),
 )
 
 //深度图使用
 Container(
   height: 230,
   width: double.infinity,
   child: DepthChart(_bids, _asks),
 )         
```
#### 3.修改样式
可在chart_style.dart里面修改图表样式

#### 4.数据处理
```dart
//接口获取数据后，计算数据
DataUtil.calculate(datas);
//更新最后一条数据
DataUtil.updateLastData(datas);
//添加数据
DataUtil.addLastData(datas,kLineEntity);
```

#### 国际化 l10n
```dart
import 'package:technixo_k_chart/generated/l10n.dart' as k_chart;
MaterialApp(
      localizationsDelegates: [
        k_chart.S.delegate//国际化
      ],
);
```

#### 5.[修改日志](https://github.com/mdat31/technixo_k_chart/blob/main/README.md)

## 问题
使用中如果有问题可以加QQ群：114563912

#### 请咖啡☕️
🙏感谢🙏

![微信](https://user-images.githubusercontent.com/20394691/102620051-61587480-4178-11eb-89e9-53686bb1c0f1.jpg)


CHANGELOG.md
## [0.0.6]
## - Feat: Add EMA calculate and drawLine

## [0.0.5+3]
## - Todo: Add extension for MainState and SecondaryState

## [0.0.5+2]
## - Todo: change text position

## [0.0.5] 
## - Todo: add text center painter

## [0.0.4] 
## - Todo: add NInterval enum and helpful extension
## - Todo: add startTime to KLineEntity
## - Todo: add example of add of update data

## [0.0.3 + 1] 
## - Todo: param model and entity to dynamic

## [0.0.3] 
## - TODO: Added k_line_model to get stream from binance api.
## - TODO: required maDayList in DataUtil.calculate.
## - TODO: Added example.

## [0.0.1+1] - TODO: Minor Changes.

## [0.0.1] - TODO: Add release date.

* TODO: Describe initial release.


example/lib/main_candle.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_k_chart/entity/candle_chart/candle_chart_entity.dart';
import 'package:flutter_k_chart/flutter_k_chart.dart';
import 'package:flutter_k_chart/generated/l10n.dart' as k_chart;
import 'package:flutter_k_chart/k_chart_widget.dart';
import 'package:flutter_k_chart/utils/data_util_binance.dart';
import 'package:http/http.dart' as http;

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      // supportedLocales: [const Locale('zh', 'CN')],
      localizationsDelegates: [
        k_chart.S.delegate //国际话
      ],
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<CandleChartEntity>? datasBinance;

  // List<DepthEntity>? _bids, _asks;
  List<int> maDayList = [10, 100, 1000];

  MainState _mainState = MainState.MA;
  SecondaryState _secondaryState = SecondaryState.NONE;

  bool volHidden = true;
  bool showLoading = true;
  bool isLine = false;

  @override
  void initState() {
    super.initState();
    getData();
    // rootBundle.loadString('assets/depth.json').then((result) {
    //   final parseJson = json.decode(result);
    //   final tick = parseJson['tick'] as Map<String, dynamic>;
    //   final List<DepthEntity> bids = (tick['bids'] as List<dynamic>)
    //       .map<DepthEntity>(
    //           (item) => DepthEntity(item[0] as double, item[1] as double))
    //       .toList();
    //   final List<DepthEntity> asks = (tick['asks'] as List<dynamic>)
    //       .map((item) => DepthEntity(item[0] as double, item[1] as double))
    //       .toList()
    //       .cast<DepthEntity>();
    //   initDepth(bids, asks);
    // });
  }

  // void initDepth(List<DepthEntity>? bids, List<DepthEntity>? asks) {
  //   if (bids == null || asks == null || bids.isEmpty || asks.isEmpty) return;
  //   _bids = [];
  //   _asks = [];
  //   double amount = 0.0;
  //   bids.sort((left, right) => left.price.compareTo(right.price));
  //   for (final item in bids.reversed) {
  //     amount += item.amount;
  //     item.amount = amount;
  //     _bids!.insert(0, item);
  //   }
  //
  //   amount = 0.0;
  //   asks.sort((left, right) => left.price.compareTo(right.price));
  //   for (final item in asks) {
  //     amount += item.amount;
  //     item.amount = amount;
  //     _asks!.add(item);
  //   }
  //   setState(() {});
  // }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: ListView(
          shrinkWrap: true,
          physics: const NeverScrollableScrollPhysics(),
          children: <Widget>[
            Stack(children: <Widget>[
              Container(
                height: 450,
                width: double.infinity,
                child: CandleChartWidget(
                  datasBinance,
                  isLine: isLine,
                  volHidden: volHidden,
                  mainState: _mainState,
                  secondaryState: _secondaryState,
                  maDayList: maDayList,
                  timeFormat: TimeFormat.YEAR_MONTH_DAY,
                ),
              ),
              if (showLoading)
                Container(
                    width: double.infinity,
                    height: 450,
                    alignment: Alignment.center,
                    child: const CircularProgressIndicator()),
            ]),
            buildButtons(),
            // if (_bids != null && _asks != null)
            //   Container(
            //     height: 230,
            //     width: double.infinity,
            //     child: DepthChart(_bids!, _asks!),
            //   )
          ],
        ),
      ),
    );
  }

  Widget buildButtons() {
    return Wrap(
      alignment: WrapAlignment.spaceEvenly,
      children: <Widget>[
        button('kLine', onPressed: () => isLine = !isLine),
        button('volHidden', onPressed: () => volHidden = !volHidden),
        button('MA',
            onPressed: () => _mainState =
                (_mainState == MainState.NONE) ? MainState.MA : MainState.NONE),
        button('BOLL',
            onPressed: () => _mainState = (_mainState == MainState.NONE)
                ? MainState.BOLL
                : MainState.NONE),
        button('MACD',
            onPressed: () => _secondaryState =
                (_secondaryState == SecondaryState.NONE)
                    ? SecondaryState.MACD
                    : SecondaryState.NONE),
        button('KDJ',
            onPressed: () => _secondaryState =
                (_secondaryState == SecondaryState.NONE)
                    ? SecondaryState.KDJ
                    : SecondaryState.NONE),
        button('RSI',
            onPressed: () => _secondaryState =
                (_secondaryState == SecondaryState.NONE)
                    ? SecondaryState.RSI
                    : SecondaryState.NONE),
        button('WR',
            onPressed: () => _secondaryState =
                (_secondaryState == SecondaryState.NONE)
                    ? SecondaryState.WR
                    : SecondaryState.NONE),
        button('CCI',
            onPressed: () => _secondaryState =
                (_secondaryState == SecondaryState.NONE)
                    ? SecondaryState.CCI
                    : SecondaryState.NONE),
        button('update', onPressed: () {
          //更新最后一条数据
          // datas!.last.close =
          //     datas!.last.close + (Random().nextInt(100) - 50).toDouble();
          // datas!.last.high = max(datas!.last.high, datas!.last.close);
          // datas!.last.low = min(datas!.last.low, datas!.last.close);
          // DataUtilBinance.updateLastData(datasBinance, maDayList);
        }),
        button('addData', onPressed: () {
          // Sao chép một đối tượng, sửa đổi dữ liệu
          // final kLineEntity = KLineEntity.fromJson(datas!.last.toJson());
          // kLineEntity.id = kLineEntity.id! + 60 * 60 * 24;
          // kLineEntity.open = kLineEntity.close;
          // kLineEntity.close =
          //     kLineEntity.close + (Random().nextInt(100) - 50).toDouble();
          // datas!.last.high = max(datas!.last.high, datas!.last.close);
          // datas!.last.low = min(datas!.last.low, datas!.last.close);
          // DataUtilBinance.addLastData(datasBinance, kLineEntity, maDayList);
        }),
        button('1month', onPressed: () async {
//              showLoading = true;
//              setState(() {});
          //getData('1mon');
          final String result = await rootBundle.loadString('assets/kmon.json');
          final parseJson = json.decode(result) as Map<String, dynamic>;
          final list = parseJson['data'] as List<dynamic>;
          // datas = list
          //     .map((item) => KLineEntity.fromJson(item as Map<String, dynamic>))
          //     .toList()
          //     .reversed
          //     .toList()
          //     .cast<KLineEntity>();
          DataUtilBinance.calculate(datasBinance, maDayList);
        }),
        button('1Day', onPressed: () {
          showLoading = true;
          setState(() {});
          getData(period: '1day');
        }),
      ],
    );
  }

  Widget button(String text, {VoidCallback? onPressed}) {
    return TextButton(
      onPressed: () {
        if (onPressed != null) {
          onPressed();
          setState(() {});
        }
      },
      child: Text(text),
      style: TextButton.styleFrom(
        primary: Colors.white,
        minimumSize: const Size(88, 44),
        padding: const EdgeInsets.symmetric(horizontal: 16.0),
        shape: const RoundedRectangleBorder(
          borderRadius: BorderRadius.all(Radius.circular(2.0)),
        ),
        backgroundColor: Colors.blue,
      ),
    );
  }

  void getData({String? symbol, String? period}) async {
    late String result;
    try {
      result = await getIPAddress(symbol: symbol, period: period);
    } catch (e) {
      print('Không lấy được dữ liệu, lấy dữ liệu cục bộ');
      result = await rootBundle.loadString('assets/kline.json');
    } finally {
      final results = json.decode(result) as List<dynamic>;
      datasBinance = results
          .map((list) {
            return CandleChartEntity.fromApi(list);
          })
          .toList()
          .cast<CandleChartEntity>();
      // debugPrint('### $datasBinance');
      DataUtilBinance.calculate(datasBinance, maDayList);
      showLoading = false;
      setState(() {});
    }
  }

  Future<String> getIPAddress({String? symbol, String? period}) async {
    final houbi =
        'https://api.huobi.br.com/market/history/kline?period=${period ?? '1day'}&size=300&symbol=btcusdt';
    final binance =
        'https://fapi.binance.com/fapi/v1/klines?symbol=${symbol ?? 'btcusdt'}&interval=${period ?? '1d'}';
    String result;
    final response =
        await http.get(Uri.parse(binance)).timeout(const Duration(seconds: 7));
    if (response.statusCode == 200) {
      result = response.body;
    } else {
      return Future.error('Nhận thất bại');
    }

    return result;
  }

  void getDataHb({String? symbol, String? period}) async {
    late String result;
    try {
      result = await getIPAddress(symbol: symbol, period: period);
    } catch (e) {
      print('Không lấy được dữ liệu, lấy dữ liệu cục bộ');
      result = await rootBundle.loadString('assets/kline.json');
    } finally {
      final parseJson = json.decode(result) as Map<String, dynamic>;
      final list = parseJson['data'] as List<dynamic>;
      // datas = list
      //     .map((item) => KLineEntity.fromJson(item as Map<String, dynamic>))
      //     .toList()
      //     .reversed
      //     .toList()
      //     .cast<KLineEntity>();
      // DataUtil.calculate(datas, maDayList);
      showLoading = false;
      setState(() {});
    }
  }

  Future<String> getIPAddressHb({String? symbol, String? period}) async {
    //Huobi api, cần khắc phục bức tường
    final houbi =
        'https://api.huobi.br.com/market/history/kline?period=${period ?? '1day'}&size=300&symbol=btcusdt';
    final binance =
        'https://fapi.binance.com/fapi/v1/klines?symbol=${symbol ?? 'btcusdt'}&interval=${period ?? '1M'}';
    String result;
    final response =
        await http.get(Uri.parse(houbi)).timeout(const Duration(seconds: 7));
    if (response.statusCode == 200) {
      result = response.body;
    } else {
      return Future.error('Nhận thất bại');
    }
    return result;
  }
}
